BUG: Check for null values when infering excel in read_clipboard

saucoide opened the pull request:
GH41108
Stripping whitespace on read_clipboard causes data copied from
excel to lose the tab separator between the first column and
the nexts, shifting the data into the wrong column

This just removes the .lstrip() except for the first line
containing headers, to avoid breaking header-less index in csv
files

- [x] closes #41108
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

saucoide committed:
BUG: Check for null values when infering excel in read_clipboard

GH41108
Stripping whitespace on read_clipboard causes data copied from
excel to lose the tab separator between the first column and
the nexts, shifting the data into the wrong column

This just removes the .lstrip() except for the first line
containing headers, to avoid breaking header-less index in csv
files

pep8speaks commented:
Hello @saucoide! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/tests/io/test_clipboard.py`](https://github.com/pandas-dev/pandas/blob/27234a21becc77087facd2ededc27258efb120bd/pandas/tests/io/test_clipboard.py):

> [Line 228:13](https://github.com/pandas-dev/pandas/blob/27234a21becc77087facd2ededc27258efb120bd/pandas/tests/io/test_clipboard.py#L228): [E101](https://duckduckgo.com/?q=pep8%20E101) indentation contains mixed spaces and tabs
> [Line 228:13](https://github.com/pandas-dev/pandas/blob/27234a21becc77087facd2ededc27258efb120bd/pandas/tests/io/test_clipboard.py#L228): [W191](https://duckduckgo.com/?q=pep8%20W191) indentation contains tabs

phofl commented on pandas/tests/io/test_clipboard.py:
```diff
@@ -220,6 +220,21 @@ def test_read_clipboard_infer_excel(self, request, mock_clipboard):
         # excel data is parsed correctly
         assert df.iloc[1][1] == "Harry Carney"
 
+        # a null on the first column works
```
Please make a separate test and link the issue

saucoide replied:
Hi @phofl, just updated it, let me know if I should change something else, thanks

phofl commented:
Please add whatsnew and fix linting issue

saucoide committed:
separated test & added whatsnew

phofl commented on pandas/tests/io/test_clipboard.py:
```diff
@@ -243,6 +243,17 @@ def test_read_clipboard_infer_excel(self, request, mock_clipboard):
 
         tm.assert_frame_equal(res, exp)
 
+    # Tests that nulls on the first column do not trip infering excel format
```
Please remove

saucoide replied:
done

phofl commented on pandas/tests/io/test_clipboard.py:
```diff
@@ -243,6 +243,17 @@ def test_read_clipboard_infer_excel(self, request, mock_clipboard):
+        df = read_clipboard()
+
+        # excel data is parsed correctly
+        assert df.iloc[1][1] == "blue"
```
please test whole frame, not only one cell.

saucoide replied:
just added an extra check on the whole frame

phofl replied:
Please use tm.assert_frame_equal,

You can remove the iloc check

saucoide replied:
sorry didn't know about that, just changed it, thanks

saucoide committed:
test compares the whole dataframe & reformatted

saucoide committed:
change test to use tm.assert_frame_equal

rhshadrach commented on pandas/io/clipboards.py:
```diff
@@ -58,7 +58,9 @@ def read_clipboard(sep=r"\s+", **kwargs):  # pragma: no cover
     # 1  3  4
 
-    counts = {x.lstrip().count("\t") for x in lines}
+    counts = {
+        x.count("\t") if i > 0 else x.lstrip().count("\t") for i, x in enumerate(lines)
+    }
```
Does this solve the issue when columns are a MultiIndex? Would using `.lstrip(" ")` be better?

saucoide replied:
You are right it doesn't work well with MultiIndex, in general it doesnt work when you have an Index at all

The problem happens when the separator is "\s+" and there are nulls in the columns/index

There was this comment in the code and i didn't want to break the behavior, so to keep it as it was, if there is a index the separator remains as `\s+`
```
# Need to remove leading white space, since read_csv
# accepts:
#    a  b
# 0  1  2
# 1  3  4
```

But that means the values still end up in the wrong column for those cases

 `.lstrip(" ")` or just removing the .lstrip completely as in:
```python
counts = {x.count("\t") for x in lines}
```
would correctly infer excel and set `sep="\t"`, but then it does not set the indexes by default, and instead they show up as `Unnamed: 0` columns, I still think that's preferable to reading the values in the wrong column though

but for example it has the consequence of turning this:
```
		a	b
A	0	1	2
A	1	3	4
A	3	5	6

```
```python
In [3]: pandas.read_clipboard()
Out[3]: 
     a  b
A 0  1  2
  1  3  4
  3  5  6
```

into this:
```python
In [8]: pandas.read_clipboard()
Out[8]: 
  Unnamed: 0  Unnamed: 1  a  b
0          A           0  1  2
1          A           1  3  4
2          A           3  5  6
```

on the other hand, if there are nulls, or merged cells as the index, the current behavior is going to give you the wrong results
```
		a	b
A	0	1	2
A	1		4
A	3	5	6

```
```python
In [11]: pandas.read_clipboard()
Out[11]: 
     a    b
A 0  1  2.0
  1  4  NaN
  3  5  6.0
```

I really dont know what's better considering this is an interactive function

rhshadrach replied:
> You are right it doesn't work well with MultiIndex, in general it doesnt work when you have an Index at all

Is this remark about this PR, or is it about the current state even in master?

> I really dont know what's better considering this is an interactive function

In general, my opinion is that it is better to me to have stability. In other words, don't fix one bug at the expense of causing another. Gets tricky when one is viewed as serious and the other as minor.

saucoide replied:
> Is this remark about this PR, or is it about the current state even in master?

Both, when there are Indexes my changed did not change the behavior, but also did not fix the problem with the columns shifting left

I've just made some changes that I think should make it work

rhshadrach reviewed (changes requested):
Thanks for the PR! Question on the fix.

saucoide committed:
Check for index columns and pass through the kwarg to read_csv

Use the number of leading tab's in the first row when we are
copying from excel as the value for `index_col`

saucoide committed:
removing print

rhshadrach commented on pandas/io/clipboards.py:
```diff
@@ -58,9 +58,14 @@ def read_clipboard(sep=r"\s+", **kwargs):  # pragma: no cover
+        # to account for index columns
+        index_length = len(lines[0]) - len(lines[0].lstrip(" \t"))
+        if index_length != 0:
+            kwargs.setdefault("index_col", [0, index_length - 1])
```
I think this won't be correct for a MultiIndex with 3 or more levels. In such a case, don't we want list(range(index_length))?

rhshadrach commented on pandas/tests/io/test_clipboard.py:
```diff
@@ -243,6 +244,34 @@ def test_read_clipboard_infer_excel(self, request, mock_clipboard):
+
+    def test_infer_excel_with_multiindex(self, request, mock_clipboard):
+        # GH41108
+        text = "\t\tcol1\tcol2\nA\t0\t1\tred\nA\t1\t\tblue\nB\t0\t2\tgreen"
```
It can be that the second row does not have A as well. Removing A here, I think the test fails, but it also fails on master. Okay to raise an issue about clipboard supporting MultiIndex better (if one doesn't already exist)

saucoide replied:
It doesn't seem to be failing when I try it with this:

```
\t\tcol1\tcol2
A\t0\t1\tred
\t1\t\tblue
B\t0\t2\tgreen
```

Is this what you mean? maybe i'm not understanding it correctly

rhshadrach commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -796,6 +796,7 @@ I/O
 - Bug in :func:`read_excel` dropping empty values from single-column spreadsheets (:issue:`39808`)
 - Bug in :meth:`DataFrame.to_string` misplacing the truncation column when ``index=False`` (:issue:`40907`)
 - Bug in :func:`read_orc` always raising ``AttributeError`` (:issue:`40918`)
+- Bug in :func:`read_clipboard` when copying from excel and the first column contains null values (:issue:`41108`)
```
Can you include what happens in the bug, something like "would incorrectly shift values between columns"

saucoide replied:
changed it to this, would that work?

```diff
+ - Bug in :func:`read_clipboard` copying from an excel file shifts values into the wrong column if there are null values in first column (:issue:`41108`)
```

rhshadrach reviewed (changes requested):
Thanks for the update, I think the approach looks good. Some requests/comments

saucoide committed:
parametrized multiindex test & fixed the args to index_col

rhshadrach commented on pandas/tests/io/test_clipboard.py:
```diff
@@ -243,6 +243,52 @@ def test_read_clipboard_infer_excel(self, request, mock_clipboard):
+    def test_infer_excel_with_multiindex(self, request, mock_clipboard, multiindex):
+        # GH41108
+
+        # the `.replace()` is because `.dedent()` does not like the leading `\t`
+        mock_clipboard[request.node.name] = multiindex[0].replace(" ", "")
```
Can you explain this more? Where is dedent being used?

saucoide replied:
To make it a bit more readable I changed the text used to mock the clipboard from
```python
text = "\t\tcol1\tcol2\nA\t0\t1\tred\nA\t1\t\tblue\nB\t0\t2\tgreen"
```
to 
```python
               """\t\t\tcol1\tcol2
                A\t0\tTrue\t1\tred
                A\t1\tTrue\t\tblue
                B\t0\tFalse\t2\tgreen"""
```

That's similar to how it's done in some of the other tests, except now there are extra spaces/tabs because of the multiline string.   
The other tests use `.dedent()` to get rid of it, but in this case it is also deleting the leading tabs for the index which is what we are testing and i couldn't get it to work properly, so I'm instead removing the spaces directly

rhshadrach replied:
That makes sense - so maybe change the comment to something more like "Can't use dedent as it will remove the leading \t" and move to the line where this string is set above.

As of now, the comment reads to me something like "we use replace because if we didn't, when we use dedent later things will go wrong".

saucoide replied:
I think you are right and the replace was a bit weird to begin with, i've changed it now to concatenate each line, i hope that's better

rhshadrach reviewed (changes requested):
Thanks for the updates! I think looks really good - but I don't understand the use of replace in the test.

saucoide committed:
Remove .replace() and use implicit str concatenation instead

saucoide committed:
make the concatenation explicit

rhshadrach commented:
Thanks - looks good! Can you resolve the merge conflict.

rhshadrach reviewed (approved):
lgtm, @phofl - good here?

jreback commented:
thanks @saucoide
